src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpSource.cs

```diff
@@ -568,6 +500,49 @@ public void Dispose()
             _httpClientLock.Dispose();
         }
 
+        private class ThrottledResponse : IDisposable
+        {
+            private IThrottle _throttle;
+            private readonly object _throttleLock = new object();
+
+            public ThrottledResponse(IThrottle throttle, HttpResponseMessage response)
+            {
+                if (throttle == null)
+                {
+                    throw new ArgumentNullException(nameof(throttle));
+                }
+
+                if (response == null)
+                {
+                    throw new ArgumentNullException(nameof(response));
+                }
+
+                _throttle = throttle;
+                Response = response;
+            }
+            
+            public HttpResponseMessage Response { get; }
+
+            public void Dispose()
+            {
+                try
+                {
+                    Response.Dispose();
+                }
+                finally
+                {
+                    lock (_throttleLock)
+                    {
+                        if (_throttle != null)
```

Comment:
Classic lock use pattern in this case would also include extra if before lock. But that's me being pedantic :smiley_cat: